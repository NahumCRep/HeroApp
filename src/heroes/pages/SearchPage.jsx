import { useNavigate, useLocation } from 'react-router-dom'
import HeroCard from '../components/HeroCard'
import AlertBox from '../components/AlertBox'
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm'
import { getHeroesByName } from '../helpers'

const SearchPage = () => {

  const { searchText, onInputChange, onResetForm } = useForm({ searchText: '' })
  const navigate = useNavigate()
  const location = useLocation()

  const { q } = queryString.parse(location.search)
  const heroes = getHeroesByName(q)

  const handleSearchSubmit = (e) => {
    e.preventDefault()
    // if(searchText.trim().length <= 1) return;
    navigate(`?q=${searchText}`)
  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <form onSubmit={handleSearchSubmit}>
            <input
              type="text"
              placeholder="Search a hero..."
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-primary mt-2">
              Search
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          <AlertBox query={q} dataLength={heroes.length} />

          {
            heroes.map(hero => (
              <HeroCard key={hero.id} {...hero} />
            ))
          }
        </div>
      </div>
    </>
  )
}

export default SearchPage