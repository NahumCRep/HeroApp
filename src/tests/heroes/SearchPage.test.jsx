import { render, screen, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { MemoryRouter, useNavigate } from "react-router-dom";
import SearchPage from "../../heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}))

describe('Pruebas sobre el componente <SearchPage />', () => { 

    beforeEach(() => jest.clearAllMocks())
    
    test('debe de hacer match con el snapshot', () => { 
        const {container} = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )
        expect(container).toMatchSnapshot()
    })

    test('debe de mostrar a batman y el input con el queryString', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )
        // screen.debug()

        const input = screen.getByRole('textbox')
        expect(input.value).toBe('batman')

        const img = screen.getByRole('img')
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg')
        
        const alertBox = screen.getByLabelText('alert-default')
        expect(alertBox.style.display).toBe('none')
        
        const warningBox = screen.getByLabelText('alert-warning')
        expect(warningBox.style.display).toBe('none')
    })

    test('debe de mostrar el mensaje de error si no encuentra el heroe (batmar123)', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )

        const warningBox = screen.getByLabelText('alert-warning')
        expect(warningBox.style.display).toBe('')
        // screen.debug()
    })

    test('debe de llamar el navigate a la pantalla nueva', () => { 
        
        const hero = 'spiderman'

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )
        
        const input = screen.getByRole('textbox')
        fireEvent.change(input, {target: {value: hero}})

        const form = screen.getByRole('form')
        fireEvent.submit(form, input.value)

        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${hero}`)
        // screen.debug()
    })
})