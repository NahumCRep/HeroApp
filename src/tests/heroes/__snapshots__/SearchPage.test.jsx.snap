// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <SearchPage /> debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero..."
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="alert-default"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
      <div
        aria-label="alert-warning"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No Hero Found with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
